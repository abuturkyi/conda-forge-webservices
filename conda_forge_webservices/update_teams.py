import github
import os
import re
import logging

from conda_smithy.github import configure_github_team
import textwrap
from functools import cache

from ruamel.yaml import YAML
from conda_forge_webservices.tokens import get_gh_client

LOGGER = logging.getLogger("conda_forge_webservices.update_teams")

JINJA_PAT = re.compile(r"\{\{([^\{\}]*)\}\}")


def _jinja2_repl(match):
    return "${{" + match.group(1) + "}}"


def _filter_jinja2(line):
    return JINJA_PAT.sub(_jinja2_repl, line)


@cache
def get_filter_out_members():
    gh = github.Github(auth=github.Auth.Token(os.environ["GH_TOKEN"]))
    org = gh.get_organization("conda-forge")
    teams = ["staged-recipes", "help-r", "r"]
    gh_teams = list(org.get_team_by_slug(team) for team in teams)
    members = set()
    for team in gh_teams:
        members.update([m.login for m in team.get_members()])
    return members


def filter_members(members):
    out = get_filter_out_members()
    return [m for m in members if m not in out]


def get_handles(members):
    mem = ["@" + m for m in filter_members(members)]
    return ", ".join(mem)


class DummyMeta:
    def __init__(self, meta_yaml):
        parse_yml = YAML(typ="safe")
        parse_yml.indent(mapping=2, sequence=4, offset=2)
        parse_yml.width = 160
        parse_yml.allow_duplicate_keys = True
        self.meta = parse_yml.load(meta_yaml)


def get_recipe_contents(gh_repo):
    try:
        resp = gh_repo.get_contents("recipe/meta.yaml")
        return resp.decoded_content.decode("utf-8")
    except github.UnknownObjectException:
        resp = gh_repo.get_contents("recipe/recipe.yaml")
        return resp.decoded_content.decode("utf-8")


def get_recipe_dummy_meta(recipe_content):
    keep_lines = []
    skip = 0
    for line in recipe_content.splitlines():
        if line.strip().startswith("extra:"):
            skip += 1
        if skip > 0:
            keep_lines.append(_filter_jinja2(line))
    assert skip == 1, "team update failed due to > 1 'extra:' sections"
    return DummyMeta("\n".join(keep_lines))


def update_team(org_name, repo_name, commit=None):
    if not repo_name.endswith("-feedstock"):
        return

    team_name = repo_name.replace("-feedstock", "").lower()
    if team_name in [
        "core",
        "bot",
        "staged-recipes",
        "arm-arch",
        "systems",
    ] or team_name.startswith("help-"):
        return

    gh = get_gh_client()
    org = gh.get_organization(org_name)
    gh_repo = org.get_repo(repo_name)

    recipe_content = get_recipe_contents(gh_repo)
    meta = get_recipe_dummy_meta(recipe_content)

    (
        current_maintainers,
        prev_maintainers,
        new_conda_forge_members,
    ) = configure_github_team(
        meta,
        gh_repo,
        org,
        repo_name.replace("-feedstock", ""),
        remove=True,
    )

    if commit:
        message = textwrap.dedent("""
            Hi! This is the friendly automated conda-forge-webservice.

            I updated the Github team because of this commit.
            """)
        newm = get_handles(new_conda_forge_members)
        if newm:
            message += textwrap.dedent(
                """
                - {} {} added to conda-forge. Welcome to conda-forge!
                  Go to https://github.com/orgs/conda-forge/invitation see your invitation.
            """.format(newm, "were" if newm.count(",") >= 1 else "was")  # noqa
            )

        addm = get_handles(
            current_maintainers - prev_maintainers - new_conda_forge_members
        )
        if addm:
            message += textwrap.dedent(
                """
                - {} {} added to this feedstock maintenance team.
            """.format(addm, "were" if addm.count(",") >= 1 else "was")
            )

        if addm or newm:
            message += textwrap.dedent("""
                You should get push access to this feedstock and CI services.

                Your package won't be available for installation locally until it is built
                and synced to the anaconda.org CDN (takes 1-2 hours after the build finishes).

                Feel free to join the community on [Zulip](https://conda-forge.zulipchat.com).

                NOTE: Please make sure to not push to the repository directly.
                      Use branches in your fork for any changes and send a PR.
                      More details on this are [here](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests).
            """)  # noqa

            c = gh_repo.get_commit(commit)
            c.create_comment(message)


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser()
    parser.add_argument("org")
    parser.add_argument("repo")
    args = parser.parse_args()
    update_team(args.org, args.repo)
