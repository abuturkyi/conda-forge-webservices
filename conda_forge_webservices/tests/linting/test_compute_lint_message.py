import shutil
import textwrap
from pathlib import Path

from inline_snapshot import snapshot

from conda_forge_webservices.linting import compute_lint_message, lint_all_recipes


def data_folder():
    return Path(__file__).parent / "data"


def test_skip_ci_recipe():
    lint = compute_lint_message("conda-forge", "conda-forge-webservices", 58)
    assert lint is None


def test_skip_lint_recipe():
    lint = compute_lint_message("conda-forge", "conda-forge-webservices", 59)
    assert lint is None


def test_ci_skip_recipe():
    lint = compute_lint_message("conda-forge", "conda-forge-webservices", 65)
    assert lint is None


def test_lint_skip_recipe():
    lint = compute_lint_message("conda-forge", "conda-forge-webservices", 66)
    assert lint is None


def test_good_recipe():
    # a message similar to this comes out
    """
    Hi! This is the friendly automated conda-forge-linting service.

    I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/good_recipe/meta.yaml```) and found it was in an excellent condition.

    """  # noqa

    lint = compute_lint_message("conda-forge", "conda-forge-webservices", 16)
    assert lint is not None, lint["message"]
    assert "found it was in an excellent condition." in lint["message"], lint["message"]


def test_ok_recipe_above_good_recipe():
    expected_message = textwrap.dedent("""
    Hi! This is the friendly automated conda-forge-linting service.

    I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```, ```recipes/recipe/meta.yaml```) and found it was in an excellent condition.

    """)  # noqa

    lint = compute_lint_message("conda-forge", "conda-forge-webservices", 54)
    assert expected_message == lint["message"]


def test_ok_recipe_beside_good_recipe():
    expected_message = textwrap.dedent("""
    Hi! This is the friendly automated conda-forge-linting service.

    I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```, ```recipes/recipe/meta.yaml```) and found it was in an excellent condition.

    """)  # noqa

    lint = compute_lint_message("conda-forge", "conda-forge-webservices", 62)
    assert expected_message == lint["message"]


def test_ok_recipe_above_ignored_good_recipe():
    expected_message = textwrap.dedent("""
    Hi! This is the friendly automated conda-forge-linting service.

    I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

    """)  # noqa

    lint = compute_lint_message("conda-forge", "conda-forge-webservices", 54, True)
    assert expected_message == lint["message"]


def test_ok_recipe_beside_ignored_good_recipe():
    expected_message = textwrap.dedent("""
    Hi! This is the friendly automated conda-forge-linting service.

    I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

    """)  # noqa

    lint = compute_lint_message("conda-forge", "conda-forge-webservices", 62, True)
    assert expected_message == lint["message"]


def test_conflict_ok_recipe():
    expected_message = textwrap.dedent("""
    Hi! This is the friendly automated conda-forge-linting service.

    I was trying to look for recipes to lint for you, but it appears we have a merge conflict.
    Please try to merge or rebase with the base branch to resolve this conflict.

    Please ping the 'conda-forge/core' team (using the @ notation in a comment) if you believe this is a bug.
    """)  # noqa

    lint = compute_lint_message("conda-forge", "conda-forge-webservices", 56)
    assert lint is not None, lint["message"]
    assert expected_message == lint["message"]


def test_conflict_2_ok_recipe():
    expected_message = textwrap.dedent("""
    Hi! This is the friendly automated conda-forge-linting service.

    I was trying to look for recipes to lint for you, but it appears we have a merge conflict.
    Please try to merge or rebase with the base branch to resolve this conflict.

    Please ping the 'conda-forge/core' team (using the @ notation in a comment) if you believe this is a bug.
    """)  # noqa

    lint = compute_lint_message("conda-forge", "conda-forge-webservices", 57)
    assert lint is not None, lint["message"]
    assert expected_message == lint["message"]


def test_rattler_build_recipe():
    expected_message = textwrap.dedent("""
    Hi! This is the friendly automated conda-forge-linting service.

    I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/recipe.yaml```) and found it was in an excellent condition.


    I do have some suggestions for making it better though...


    For **recipe/recipe.yaml**:
    
    This is a rattler-build recipe and not yet lintable. We are working on it!
    """)  # noqa

    lint = compute_lint_message("conda-forge", "conda-forge-webservices", 632)
    assert lint is not None, lint["message"]
    assert expected_message == lint["message"]


def test_bad_recipe():
    # a message similar to this comes out
    """
    Hi! This is the friendly automated conda-forge-linting service.

    I wanted to let you know that I linted all conda-recipes in your PR (```recipes/bad_recipe/meta.yaml```) and found some lint.

    Here's what I've got...


    For **recipes/bad_recipe/meta.yaml**:

        * The home item is expected in the about section.
        * The license item is expected in the about section.
        * The summary item is expected in the about section.
        * The recipe must have some tests.
        * The recipe must have a `build/number` section.
        * There are 2 too many lines.  There should be one empty line at the end of the file.
        * Feedstock with the same name exists in conda-forge
        * Recipe maintainer "support" does not exist
    """  # noqa

    lint = compute_lint_message("conda-forge", "conda-forge-webservices", 17)
    assert lint is not None, lint["message"]
    assert "found some lint" in lint["message"], lint["message"]
    assert "The home item is expected in the about section." in lint["message"], lint[
        "message"
    ]
    assert "For **recipes/bad_recipe/meta.yaml**:" in lint["message"], lint["message"]


def test_mixed_recipe():
    # a message similar to this comes out
    """
    Hi! This is the friendly automated conda-forge-linting service.

    I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/hints_only/meta.yaml```) and found it was in an excellent condition.


    I do have some suggestions for making it better though...


    For **recipes/hints_only/meta.yaml**:

        * Whenever possible python packages should use pip. See https://conda-forge.org/docs/maintainer/adding_pkgs.html#use-pip
    """  # noqa

    lint = compute_lint_message("conda-forge", "conda-forge-webservices", 217)
    assert lint is not None, lint["message"]
    assert (
        "I do have some suggestions for making it better though" in lint["message"]
    ), lint["message"]  # noqa


def test_no_recipe():
    expected_message = textwrap.dedent("""
    Hi! This is the friendly automated conda-forge-linting service.

    I was trying to look for recipes to lint for you, but couldn't find any.
    Please ping the 'conda-forge/core' team (using the @ notation in a comment) if you believe this is a bug.
    """)  # noqa

    lint = compute_lint_message("conda-forge", "conda-forge-webservices", 523)
    assert lint is not None, lint["message"]
    assert expected_message == lint["message"]


def test_closed_pr():
    lint = compute_lint_message("conda-forge", "conda-forge-webservices", 52)
    assert lint is None


def test_new_recipe(tmp_path):
    recipe_file = tmp_path / "recipe" / "recipe.yaml"
    recipe_file.parent.mkdir(parents=True)
    shutil.copy(data_folder() / "recipe.yaml", recipe_file)

    message = lint_all_recipes(Path(tmp_path), [])
    assert message == snapshot(
        (
            """\

Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/recipe.yaml```) and found it was in an excellent condition.


I do have some suggestions for making it better though...


For **recipe/recipe.yaml**:

This is a rattler-build recipe and not yet lintable. We are working on it!
""",  # noqa
            "mixed",
        )
    )
