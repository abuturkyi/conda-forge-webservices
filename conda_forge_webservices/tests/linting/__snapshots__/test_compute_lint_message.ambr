# serializer version: 1
# name: test_new_recipe[rattler-build-good-lint]
  tuple(
    '''
      
      Hi! This is the friendly automated conda-forge-linting service.
      
      I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/recipe.yaml```) and found it was in an excellent condition.
      
      
      I do have some suggestions for making it better though...
      
      
      For **recipe/recipe.yaml**:
      
      This is a rattler-build recipe and not yet lintable. We are working on it!
  
    ''',
    'mixed',
  )
# ---
